TopMovies/API/GraphQL/MovieAPIGraphQL.swift

```diff
@@ -0,0 +1,126 @@
+              .map { request -> CategoryDTOWrapper in
+                .init(id: request.rawValue,
+                      title: CategoryDTO.name(by: request),
+                      fragment: fragment,
+                      request: request)
+              }
+          }
+      )
+    }
+  }
+  
+  // MARK: - Category
+  func category(request: MovieCategoryRequest,
+                page: String?,
+                categoryResult: @escaping (Result<CategoryDTOWrapper, Error>) -> Void) {
+    switch request {
+    case .nowPlaying:
+      apollo.fetch(query: NowPlayingCategoryQuery(cursor: page))
+      { [unowned self] result in
+        categoryResult(
+          result
+            .map(\.data)
+            .map(dataResult(_:))
+            .flatMap { $0 }
+            .map(\.movies.nowPlaying.fragments.movieCategoryFragment)
+            .map { categoryWrapper(by: $0, and: request) }
+        )
+      }
+    case .popular:
+      apollo.fetch(query: PopularCategoryQuery(cursor: page))
+      { [unowned self] result in
+        categoryResult(
+          result
+            .map(\.data)
+            .map(dataResult(_:))
+            .flatMap { $0 }
+            .map(\.movies.popular.fragments.movieCategoryFragment)
+            .map { categoryWrapper(by: $0, and: request) }
+        )
+      }
+    case .topRated:
+      apollo.fetch(query: TopRatedCategoryQuery(cursor: page))
+      { [unowned self] result in
+        categoryResult(
+          result
+            .map(\.data)
+            .map(dataResult(_:))
+            .flatMap { $0 }
+            .map(\.movies.topRated.fragments.movieCategoryFragment)
+            .map { categoryWrapper(by: $0, and: request) }
+        )
+      }
+    case .upcoming:
+      apollo.fetch(query: UpcomingCategoryQuery(cursor: page))
+      { [unowned self] result in
+        categoryResult(
+          result
+            .map(\.data)
+            .map(dataResult(_:))
+            .flatMap { $0 }
+            .map(\.movies.upcoming.fragments.movieCategoryFragment)
+            .map { categoryWrapper(by: $0, and: request) }
+        )
+      }
+    }
+  }
+  private func categoryWrapper(by fragment: MovieCategoryFragment,
+                               and request: MovieCategoryRequest) -> CategoryDTOWrapper {
+    .init(id: request.rawValue,
+          title: CategoryDTO.name(by: request),
+          fragment: fragment)
+  }
+  
+  // MARK: - Movie
+  func movie(id: MoviePreview.ID,
+             movieResult: @escaping (Result<MovieDTOWrapper, Error>) -> Void) {
+    guard let id = Int(id.value) else { return }
+    apollo.fetch(query: MovieQuery(id: id)) { [unowned self] result in
+      movieResult(
+        result
+          .map(\.data)
+          .map(dataResult(_:))
+          .flatMap { $0 }
+          .map(\.movies.movie.details.fragments.movieDetailsFragment)
+          .map(MovieDTOWrapper.init(fragment: ))
+      )
+    }
+  }
+  
+  private func dataResult<T>(_ data: T?) -> Result<T, Error> {
```

Comment:
I suppose that each DTO from apollo is covered with some protocol, that guarantees that it has `data` property. So that you could avoid writing
```swift        
result
    .map(\.data)
    .map(attemptParse(_:))
    .flatMap { $0 }
```
And make it shorter:
```swift        
result
    .flatMap(GraphQLResult.successfulResult)
```
Here's how it could look like:
```swift
extension MovieAPIGraphQL {
  struct AllApolloErrors: Swift.Error {
    let errors: [GraphQLError]?
  }
}

extension GraphQLResult {
  static func successfullResult(result: GraphQLResult) -> Result<Data, AnyError> {
    guard let data = result.data else {
      return .failure(AnyError(MovieAPIGraphQL.AllApolloErrors(errors: result.errors)))
    }
    return .success(data)
  }
}

```

Even though GraphQLResult is generic, you can write and extension for it and attempt data extraction.

Reply:
Wow, that's perfect, thanks !